src/vsr.zig

```diff
@@ -1571,6 +1571,52 @@ pub const Checkpoint = struct {
         assert(constants.journal_slot_count > constants.lsm_batch_multiple);
         assert(constants.journal_slot_count % constants.lsm_batch_multiple == 0);
     }
+    /// Checkpoint identifier.
+    id: u128,
+    /// The op_checkpoint() that corresponds to the checkpoint id.
+    op: u64,
+
+    pub const Quorum = struct {
+        /// The latest known checkpoint identifier from every *other* replica.
+        checkpoints: [constants.replicas_max]?Checkpoint =
+            [_]?Checkpoint{null} ** constants.replicas_max,
+
+        pub fn replace(
+            quorum: *Quorum,
+            replica: u8,
+            checkpoint: *const Checkpoint,
+        ) bool {
+            if (quorum.checkpoints[replica]) |checkpoint_existing| {
+                // Ignore old candidate.
+                if (checkpoint.op < checkpoint_existing.op) {
+                    return false;
+                }
+
+                // Ignore repeat candidate.
+                if (checkpoint.op == checkpoint_existing.op and
+                    checkpoint.id == checkpoint_existing.id)
+                {
+                    return false;
+                }
+            }
+            quorum.checkpoints[replica] = checkpoint.*;
+            return true;
+        }
+
+        pub fn count(quorum: *const Quorum, checkpoint: *const Checkpoint) usize {
+            var matching: usize = 0;
+            for (quorum.checkpoints) |checkpoint_existing| {
+                if (checkpoint_existing != null and
+                    checkpoint_existing.?.op == checkpoint.op and
+                    checkpoint_existing.?.id == checkpoint.id)
+                {
+                    assert(std.meta.eql(checkpoint.*, checkpoint_existing.?));
+                    matching += 1;
+                }
+            }
+            return matching;
```

Comment:
Should be be able to assert that `matching > 0`?

Reply:
Requiring `matching > 0` requires us to use `replace()](https://github.com/tigerbeetle/tigerbeetle/pull/2104/files#diff-e95e4dc6c1aae349e798cfeed22ad01f830cce0e654e6b4b765b3e7b0c677b6aR9511-R9514) in multiple places (commit_op_checkpoint_superblock_callback, sync_superblock_update_callback, open) to make sure that we set our own checkpoint in checkpoint_from_all_replicas every time it changes. That seemed a bit error prone so I decided to use [if (matching >= self.quorum_replication - 1)](https://github.com/tigerbeetle/tigerbeetle/pull/2104/files#diff-e95e4dc6c1aae349e798cfeed22ad01f830cce0e654e6b4b765b3e7b0c677b6aR9522-R9523) instead, assuming we already have op_checkpoint() durable.

Thoughts?

Reply:
Yeah, it seems that its easy to _not_ include us here (but than, at the call-site, we should assert that our own checkpoint is `null`.